### PR TITLE
Added FuzzySelection type interaction, updated selection prompts to have multiline prompts, added DateTimeSelector

### PR DESCRIPTION
**Pull request by slaterb1**
This interaction is similar to `Selection` but allows the user to type in characters to match against to filter to selections. For simplicity, the selected option is returned instead of it's position (that is relative depending on the number of items in the reduced set).

Moving the cursor through the selections is done via the arrow keys (same as Selection), but vim bindings (hjkl) now only effect filtering the words. 

### Added further enhancements on my fork
Now `Select`, `Checkboxes`, and `FuzzySelect` can display selections that span multiple lines and have a custom set offset.

`FuzzySelect` now includes ability to use casing in matches.

### Added DateTimeSelector on my fork
This is a new interactions that allows you to select a date by updating a displayed datetime using LEFT/RIGHT/h/l to move between positions, UP/DOWN/j/k to increment/decrement values by 1, or use numerals to set the date (if valid date) in the position of the cursor.
- default time in now with zeroed hour, min, sec
- default time can be set to anything
- weekday can be turned off
- prompt can be ignored
- can specify if you want to select a date, time, or datetime
- added examples and tests

### Bug Fix
- corrected term.size() to specify the window height not with for pagination

### CODE REVIEW

**Comment by mitsuhiko** [comment]
I played with it and it looks nice but the input for the fuzzy search is not shown anywhere and the redraw is quite flickery. Do you think this could be fixed?

**Comment by slaterb1** [comment]
@mitsuhiko I added the ability to display the `search_match` string that is being filtered against. If you build the example using --release it does not flicker. I was worried about it too when I initially build it.

**Comment by pksunkara** [comment]
I am willing to merge this now if you can rebase. Thanks.

**Comment by slaterb1** [comment]
Updated and ready to go, thanks for taking a look

**Comment by slaterb1** [comment]
@pksunkara hold up on the merge, I'm about to fix a paging bug I just discovered in all `src/select.rs`. It needs to be using `term.size().1` not `term.size().0`. terminal_size returns `{Width, Height}` not `{Height, Width}`

**Comment by slaterb1** [comment]
Done, and comment is updated.

File: src/select.rs [hunk 1]
```diff
@@ -24,6 +26,22 @@ pub struct Checkboxes<'a> {
     clear: bool,
     theme: &'a dyn Theme,
     paged: bool,
+    offset: usize,
+    lines_per_item: usize,
```
Comment (pksunkara) [inline src/select.rs:30]: I don't agree with this `lines_per_item`. Ideally, we should calculate the str width of each item after removing ansi codes and then calculate the number of lines that item takes after text wrapping it.

File: src/select.rs [hunk 2]
```diff
@@ -464,6 +509,259 @@ impl<'a> Checkboxes<'a> {
...
+        let mut size_vec = Vec::new();
+        for items in self.items.iter().as_slice() {
+            let size = &items.len();
+            size_vec.push(size.clone());
+        }
+        loop {
+            let filtered_list: Vec<&String> = self
+                .items
+                .iter()
+                .filter(|item| {
```
Comment (pksunkara) [inline src/select.rs:655]: Fuzzy selection means you something different. You don't use `find` for this. For example, `cml` should be able to match `camel`.

Comment (slaterb1) [reply src/select.rs:655]: We are currently using "fuzzy" selection to match across strings (i.e. `ch a` would match `match across`). I agree that the terminology is more like "AutocompleteSelection" but we are not changing the naming as we are using my fork in production.

File: src/select.rs (hunk 1 above)
Comment (slaterb1) [reply src/select.rs:30]: Clearly you didn't understand the intention here. You case would account for having variable number of items per page depending on the sizes of all of the items. I was using this for real examples of selecting from multi-line items:
Title:
Author:
Description:

**Comment by slaterb1** [comment]
Thanks for the feedback, despite it taking 7 months to get. We have some disagreements in terms of logic on functions, so I'll take my fork in my own direction. At least I'll be able to update the code to modern Rust styling/conventions.

We also have differences in terms of what `absolutely broken` is. You should look into paging, as the current code doesn't actually work properly, but you guys don't capture that in any of your examples. Maybe I'll open up some BUG tickets next year.

File: src/select.rs (hunk 1 above)
Comment (pksunkara) [reply src/select.rs:30]: My case would still cover the case you wanted.

**Comment by pksunkara** [comment]
Please don't hesitate to open bug tickets

**Comment by pksunkara** [comment]
And the multiline example is definitely absolutely broken because I couldn't interact with it all.

**Comment by slaterb1** [comment]
multiline_selection

**Comment by pksunkara** [comment]
I am on mac default terminal and it wasn't working. But anyway thanks for your effort. If you don't plan to PR, I will slowly port the relevant code from your fork while making sure it is correct.

Currently, I am working on fixing a few interoperability issues on windows for console library.

**Comment by slaterb1** [comment]
not going to fight it, not going to PR, but the gif was ubuntu default and I just ran it on mac default (work machine) and it still worked exactly the same.

You might want to look into the issue with your terminal, it could produce other false negatives (especially since you are furthering this library) or do yourself a favour and update to iterm2 :)

**Pull request closed without merging**